lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/BinarySpec.hs

```diff
@@ -203,3 +237,123 @@ spec = do
             let msg = "Address has unexpected length 1: Address {unAddress = \"0\"}"
             evaluate (runPut $ putAddress $ Address "0")
                 `shouldThrow` userException msg
+
+        it "decode (encode address) === address" $ property $ \addr ->
+            let
+                bytes = (runPut $ putAddress addr)
+                _hex = convertToBase @ByteString @ByteString Base16 (BL.toStrict bytes)
+            in
+                counterexample ("bytes: " <> show bytes) $ do
+                let addr' = try' $ runGet getAddress bytes
+                addr' === (Right addr)
+
+        it "decode (encode tx) === (validate tx)" $ property $ \(SignedTx signedTx) ->
+            let
+                bytes = try' (runPut $ putSignedTx signedTx)
+                hex = convertToBase @ByteString @ByteString Base16 . BL.toStrict <$> bytes
+                numberOfIns = length . inputs . fst $ signedTx
+                numberOfOuts = length . outputs . fst $ signedTx
+            in
+                counterexample ("bytes: " <> show hex) $
+                tabulate "number of inputs" [show numberOfIns] $
+                counterexample (show (numberOfIns, numberOfOuts)) $ do
+                let tx' = try' . unMessage . runGet getMessage =<< bytes
+                labelTx signedTx $ tx' === validate signedTx
+  where
+    unMessage :: Message -> (Tx, [TxWitness])
+    unMessage m = case m of
+        Transaction stx -> stx
+        _ -> error "expected a Transaction message"
+
+    try' :: a -> Either String a
+    try' = unsafePerformIO . wrap
+
+    wrap :: a -> IO (Either String a)
+    wrap = fmap (either (Left . show) Right) . (try @SomeException) . evaluate
+
+    labelTx stx = label (either id (const "Right _") (validate stx))
+
+    validate :: (Tx, [TxWitness]) -> Either String (Tx, [TxWitness])
+    validate stx@(Tx ins outs, _wits) = const stx <$> do
+        unless (length ins < 256) $
+            Left "number of inputs must be lower than 256"
+        unless (length outs < 256) $
+            Left "number of outputs must be lower than 256"
+        forM_ ins $ \(TxIn _tx ix, _coin) ->
+            unless (ix < 256) $ Left "input index must be < 256"
```

Comment:
Seems like we are re-encoding code source logic into tests which is a sign that either of them are wrong. If the source can't be easily tested, that's a problem. Or maybe, the test isn't testing at the right level of abstraction.

Reply:
Yes! I had a thought about remaking this function to
```haskell
validateSignedTx :: (Tx, [TxWitness]) -> Either ErrSignedTxValidation (Tx, [TxWitness])
```
, moving it and calling it in `mkStdTx`. We could test that `decode (encode signedTx)` throws iff `validateSignedTx` returns `Left`.

Although then it would be tempting to have jormungandr-specific types:

```haskell
data SignedTxIn = SignedTxIn Address Word8 Coin TxWitness

data SignedTx = SignedTx
    { signedInputs :: [SignedTxIn]
    , outputs :: [TxOut]
```
and make the encoders work with them… 🤔

Reply:
I am not sure this function should even exist (cf comments on the other PR). Instead:

- generators for property-based tests should generate valid values
- there should be some "negative tests" showing that, for invalid transaction types, the serialization fails.

Reply:
Upon further reflection, there are two properties I find important:
```haskell
-- Note: tx :: (Tx, [TxWitness])) below

-- **Jörmungandr-valid transactions roundtrip**
forall tx. valid tx ==>  decode (encode tx) === tx

-- **We must not be able to encode a jörmungandr-invalid transaction
-- to valid binary**. "Valid binary" means that we can decode it.
--
-- So while we don't care about general Tx round-tripping, we care about 
-- the result either 1) not being a 'Tx' or 2) being Jörmungandr-valid, in which
-- case it must be the same tx as we started with:
forall tx. case (decode (encode tx)) of
                   Right tx' -> tx' == tx
                   Left _ -> True -- we don't care
```

Sketch on the topic:
<img width="927" alt="Skärmavbild 2019-06-28 kl  14 49 29" src="https://user-images.githubusercontent.com/304423/60343373-1e1eaa80-99b4-11e9-8a80-adb211724059.png">

Reply:
And well, then we need the definition of a `[jörmungandr-]valid` transaction in one way or another. @KtorZ thoughts?

Reply:
Oh I forgot one point. By labelling the errors - either via error messages in the `Binary` module or in the `valid` function we can know that we are generating interesting enough jörmungandr-**in**valid transactions.

Reply:
Hmmm... `valid tx ==> decode (encode tx)` is an invalid use-case for the pre-condition. Just make your generator valid in the first place! If you want to test invalid transaction, then make a generator which generates invalid ones... though, this is a bit counter-productive with having a type-system in a first place.

The idea of a strongly typed system is to make sure that, having something of a given type gives us some guarantees and invariant about this type. So, by considering a `Tx`, we are assuming that it is well-formed, otherwise, we are better of having a `([TxIn], [TxOut])` and makes no assumption about the number of inputs and outputs.

We have properties checking that our coin selection preserve such invariant (and only construct valid txs). We can assume something similar for transaction coming out of Jörmungandr. However, there's nothing wrong with `expressing` those invariants in code and have some explicit assertions which would loudly fail in case they are violated. It's usually better to encode in the type system whatever we can, although here, we are really deserializing bytes from a trusted source, and there's not much we could do apart from crashing if a tx was invalid.

Reply:
I guess I was paranoid about the "strong typed system" being too weak and the safetynet failing us 🤔 

Anyhow --> https://github.com/input-output-hk/cardano-wallet/pull/476#issuecomment-507688901